### PR TITLE
DEVX-915: RBAC demo v1: Connect

### PR DESCRIPTION
**Pull request by ybyzek**

### CODE REVIEW

File: security/rbac/delta_configs/connect-avro-distributed.properties.delta
Comment (C0urante) [inline security/rbac/delta_configs/connect-avro-distributed.properties.delta:16]: I think we can (and should) retain these. The only property that we _don't_ want to set in the worker file is `sasl.jaas.config`, which provides the actual credentials that identify the service account. Since we want each connector to have its own service account, there shouldn't be a default for that.

However, everything else (`security.protocol`, `sasl.mechanism`, and `sasl.login.callback.handler`) just specifies how authentication will happen (but not user credentials), so it's fine to set those as global defaults that every connector will inherit unless explicitly overridden.

File: security/rbac/delta_configs/connect-avro-distributed.properties.delta
Comment (C0urante) [inline security/rbac/delta_configs/connect-avro-distributed.properties.delta]: OOC, do things break if this isn't specified?

File: security/rbac/delta_configs/connector-source.properties.delta
Comment (C0urante) [inline security/rbac/delta_configs/connector-source.properties.delta]: Following up on comments left on the `connect-avro-distributed.properties.delta` file--we should remove the `producer.override.sasl.mechanism` and `producer.override.sasl.login.callback.handler` configs from this file and they should instead be set in the worker file.

File: security/rbac/scripts/submit_datagen_pageviews_config_avro.sh
Comment (C0urante) [inline security/rbac/scripts/submit_datagen_pageviews_config_avro.sh]: Following up on comments left on the `connect-avro-distributed.properties.delta` file--we should remove the `producer.override.sasl.mechanism` and `producer.override.sasl.login.callback.handler` configs from this config and they should instead be set in the worker file.

File: security/rbac/scripts/submit_datagen_pageviews_config_avro.sh
Comment (C0urante) [inline security/rbac/scripts/submit_datagen_pageviews_config_avro.sh:21]: These credentials should match the ones in the `producer.override.sasl.jaas.config`, so that there's one service account to track for the connector that deals with both Kafka interactions (writing to, reading from, and creating topics) and Schema Registry interactions (writing to and reading from subjects).

File: security/rbac/scripts/enable-rbac-connect.sh
Comment (C0urante) [inline security/rbac/scripts/enable-rbac-connect.sh]: Nit: I think the language here can be tweaked. My understanding is that a [principal](https://docs.oracle.com/javase/8/docs/api/java/security/Principal.html) (at least in the Java ecosystem) is an identity like a user or a service account, so I think it might make more sense to word this like "Grant the principal <principal> the role <role> for the resource <resource>" or "Grant the principal <principal> the role <role> for the cluster <cluster>".

File: security/rbac/scripts/enable-rbac-connect.sh
Comment (C0urante) [inline security/rbac/scripts/enable-rbac-connect.sh]: This isn't necessary; connectors don't need permissions to interact with themselves.

File: security/rbac/scripts/enable-rbac-connect.sh
Comment (C0urante) [inline security/rbac/scripts/enable-rbac-connect.sh]: This isn't necessary; connectors don't need permissions to interact with themselves.

File: security/rbac/scripts/enable-rbac-connect.sh
Comment (C0urante) [inline security/rbac/scripts/enable-rbac-connect.sh]: This makes the service account for the Connect worker the ResourceOwner for the connector, which is a little strange.

Being the ResourceOwner for a connector allows someone to view its configuration, pause/resume it, create it, check its status, things like that. So that role should belong to a user, specifically, the one that issues the REST request to create the connector. Is there a user principal defined in this example?

File: security/rbac/scripts/enable-rbac-connect.sh
Comment (C0urante) [inline security/rbac/scripts/enable-rbac-connect.sh]: Should be `User:connector` instead of `User:$CLIENT` (left an in-depth explanation as a comment on the `submit_datagen_pageviews_config_avro.sh` file).

File: security/rbac/scripts/submit_datagen_pageviews_config.sh
Comment (C0urante) [inline security/rbac/scripts/submit_datagen_pageviews_config.sh]: Following up on comments left on the connect-avro-distributed.properties.delta file--we should remove the producer.override.sasl.mechanism and producer.override.sasl.login.callback.handler configs from this config and they should instead be set in the worker file.

**Review by C0urante — COMMENTED** [review]
Looks pretty good! I think there's still a bit of a disconnect on two fronts (client security configs in worker file vs. client security configs in connector file, and which principals are in play and where they should be used between the user, connectors, and the Connect worker), but the overall content is great.

**Review by tobydrake7 — COMMENTED** [review]
Yeva, this looks great. I hope to provide review comments when I go through the steps myself later today.

File: security/rbac/delta_configs/connect-avro-distributed.properties.delta
Comment (ybyzek) [reply security/rbac/delta_configs/connect-avro-distributed.properties.delta:16]: @C0urante thanks for the explanation and it makes sense.  I've modified the code, please let me know what you think.

File: security/rbac/delta_configs/connect-avro-distributed.properties.delta
Comment (ybyzek) [reply security/rbac/delta_configs/connect-avro-distributed.properties.delta]: Nope!  I'll cut it.

File: security/rbac/delta_configs/connector-source.properties.delta
Comment (ybyzek) [reply security/rbac/delta_configs/connector-source.properties.delta]: Agree

File: security/rbac/scripts/submit_datagen_pageviews_config_avro.sh
Comment (ybyzek) [reply security/rbac/scripts/submit_datagen_pageviews_config_avro.sh]: Agree

File: security/rbac/scripts/submit_datagen_pageviews_config_avro.sh
Comment (ybyzek) [reply security/rbac/scripts/submit_datagen_pageviews_config_avro.sh:21]: One SA for both the connector and SR makes logical sense, but is that the way it will always work?  It may be different, right?, depending on how SA creds are handed out?

File: security/rbac/delta_configs/connect-avro-distributed.properties.delta
Comment (C0urante) [reply security/rbac/delta_configs/connect-avro-distributed.properties.delta:16]: Looks good!

File: security/rbac/scripts/submit_datagen_pageviews_config_avro.sh
Comment (C0urante) [reply security/rbac/scripts/submit_datagen_pageviews_config_avro.sh:21]: While it's certainly possible, I don't think it's likely, and it goes against the recommended flow of one SA per connector. I think we should demonstrate the common case here, but I'll leave that up to you.

File: security/rbac/scripts/enable-rbac-connect.sh
Comment (ybyzek) [reply security/rbac/scripts/enable-rbac-connect.sh]: Validated and removed

File: security/rbac/scripts/enable-rbac-connect.sh
Comment (ybyzek) [reply security/rbac/scripts/enable-rbac-connect.sh]: Validated and removed

File: security/rbac/scripts/submit_datagen_pageviews_config_avro.sh
Comment (ybyzek) [reply security/rbac/scripts/submit_datagen_pageviews_config_avro.sh:21]: I'm changing everything from `connector` to `client`.

File: security/rbac/scripts/enable-rbac-connect.sh
Comment (ybyzek) [reply security/rbac/scripts/enable-rbac-connect.sh]: I'm changing everything from `connector` to `client`.

File: security/rbac/scripts/enable-rbac-connect.sh
Comment (ybyzek) [reply security/rbac/scripts/enable-rbac-connect.sh]: @C0urante thanks for bringing this up.  I'll pursue this separately, and it will apply to all components, not just connect

Comment (ybyzek) [reply security/rbac/scripts/enable-rbac-connect.sh]: Resolved

File: security/rbac/scripts/enable-rbac-connect.sh
Comment (framiere) [inline security/rbac/scripts/enable-rbac-connect.sh]: When the rbac demo will get bigger, we could maybe leverage associative arrays like such:

```sh
resources=(
    'Group:connect-cluster::'
    'Topics:_secrets::(for secret Registry)'
    'Group:secret-registry::(for secret Registry)')

for resource in "${resources[@]}" ; do
    RESOURCE="${resource%%::*}"
    DESCRIPTION="${resource##*::}"

    echo -e "\n# Grant the principal User:$ADMIN_CONNECT to the ResourceOwner role for $RESOURCE $DESCRIPTION"
    echo "confluent iam rolebinding create --principal User:$ADMIN_CONNECT --role ResourceOwner --resource $RESOURCE --kafka-cluster-id $KAFKA_CLUSTER_ID"
    confluent iam rolebinding create --principal User:$ADMIN_CONNECT --role ResourceOwner --resource $RESOURCE --kafka-cluster-id $KAFKA_CLUSTER_ID
done
```

Comment (ybyzek) [reply security/rbac/scripts/enable-rbac-connect.sh]: @framiere that's very cool bash mojo, but I intentionally didn't make those optimizations for the following reasons:

1. This demo is meant as a learning resource
2. For loops obscure all the required steps...all steps should be unfurled
3. Fancy bash makes it harder for bash newbies to parse
4. Optimizations should be made in a configuration tool

(In fact, I just unfurled my only for loop I had somewhere else)

**Comment by ybyzek** [comment]
I believe I addressed all comments in this PR.  Please review/approve if it looks good

**Review by tobydrake7 — APPROVED** [review]
LGTM 👍

**Pull request merged**
